File: frontend/src/lib/types/news.ts
export type Article = {
  publish_date: string;
  id: number;
  url: string;
  author: string;
  article_text: string;
  title: string;
  main_image: string;
  province: string;
  city: string;
  latitude: number;
  longitude: number;
  search_vector: string;
  rank: number;
};

export type News = {
  data: Article[];
  count: number;
  page: number;
  size: number;
  total_pages: number;
  has_next: boolean;
  has_prev: boolean;
};

export type NewsSource = {
  author: string;
  article_count: number;
};

export type ArticleTrend = {
  bucket: string;
  article_count: number;
};

export type ProvinceSummary = {
  province: string;
  article_count: number;
};

type GeoJSONGeometry = {
  type: string;
  coordinates: number[];
};

type Properties = {
  name: string;
  city: string;
  province: string;
  article_count: number;
};

export type Feature = {
  type: string;
  geometry: GeoJSONGeometry;
  properties: Properties;
};

export type CitySummary = {
  type: string;
  features: Feature[];
};

File: backend/app/models/news.py
from datetime import datetime
from enum import Enum

from sqlmodel import Field, SQLModel


class NewsBase(SQLModel):
    title: str = Field(index=True)
    author: str | None = None
    publish_date: datetime | None = Field(index=True)
    article_text: str
    url: str | None = None
    main_image: str | None = None
    city: str | None = Field(index=True)
    province: str | None = Field(index=True)
    latitude: float | None = None
    longitude: float | None = None


class NewsWithRank(NewsBase):
    id: int | None
    rank: float | None = Field(default=0.0)


class NewsPublic(SQLModel):
    data: list[NewsWithRank]
    count: int
    page: int
    size: int
    total_pages: int
    has_next: bool
    has_prev: bool


class TopNewsSource(SQLModel):
    author: str
    article_count: int


class TopNewsSourcePublic(SQLModel):
    data: list[TopNewsSource]


class Granularity(str, Enum):
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"


class NewsTrends(SQLModel):
    bucket: datetime
    article_count: int


class NewsTrendsPublic(SQLModel):
    data: list[NewsTrends]


class ProvinceSummary(SQLModel):
    province: str
    article_count: int


class ProvinceSummaryPublic(SQLModel):
    data: list[ProvinceSummary]


class GeoJSONGeometry(SQLModel):
    type: str = "Point"
    coordinates: list[float]


class CitySummaryProperties(SQLModel):
    city: str
    province: str
    article_count: int


class GeoJSONFeature(SQLModel):
    type: str = "Feature"
    geometry: GeoJSONGeometry
    properties: CitySummaryProperties


class GeoJSONFeatureCollection(SQLModel):
    type: str = "FeatureCollection"
    features: list[GeoJSONFeature]


class CitySummaryResult(SQLModel):
    city: str
    province: str
    latitude: float
    longitude: float
    article_count: int
